### PR TITLE
adds startupMonitorPodConditionController for monitoring the lifecycle of the startup monitor pod

### PR DESCRIPTION
**Pull request by p0lyn0mial**
xref: openshift/enhancements#833

### CODE REVIEW

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	corev1listers "k8s.io/client-go/listers/core/v1"
```
Comment (sttts) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:15]: order

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+		startupMonitorEnabledFn: startupMonitorEnabledFn,
+	}
+	return factory.New().WithSync(fd.sync).ResyncEvery(6*time.Minute).WithInformers(kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Informer()).ToController("StartupMonitorPodCondition", eventRecorder)
+}
+
+func (fd *startupMonitorPodConditionController) sync(_ context.Context, _ factory.SyncContext) (err error) {
+	startupPodPendingDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodPendingDegraded", Status: operatorv1.ConditionFalse}
+	startupPodContainerWaitingDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodContainerWaitingDegraded", Status: operatorv1.ConditionFalse}
+	startupPodFailedContainerDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodFailedContainerDegraded", Status: operatorv1.ConditionFalse}
+	startupPodContainerExcessiveRestartsDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodContainerExcessiveRestartsDegraded", Status: operatorv1.ConditionFalse}
```
Comment (sttts) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:48]: these are many conditions. I don't think we want one for each symptom, but rather use the reason and message to distinguish.

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+func (fd *startupMonitorPodConditionController) sync(_ context.Context, _ factory.SyncContext) (err error) {
+	startupPodDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodDegraded", Status: operatorv1.ConditionFalse}
+	startupPodContainerExcessiveRestartsDegraded := &operatorv1.OperatorCondition{Type: "StartupMonitorPodContainerExcessiveRestartsDegraded", Status: operatorv1.ConditionFalse}
+
+	defer func() {
+		if err == nil {
+			if _, _, updateError := operatorv1helpers.UpdateStatus(fd.operatorClient,
+				operatorv1helpers.UpdateConditionFn(*startupPodDegraded),
+				operatorv1helpers.UpdateConditionFn(*startupPodContainerExcessiveRestartsDegraded)); updateError != nil {
```
Comment (sttts) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:52]: do we have a variant of `UpdateConditionFn` that also remove the condition if it is not needed?

Comment (sttts) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:52]: i.e. if `startupMonitorEnabledFn()` is false

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+	withDegradedWhenInPendingPhaseFor(monitorPod, 2*time.Minute, startupPodDegraded)
+	withDegradedWhenHasFailedContainer(monitorPod, startupPodDegraded)
+	withDegradedWhenHasExcessiveRestarts(monitorPod, 2, startupPodContainerExcessiveRestartsDegraded)
+
+	return nil
+}
+
+func withDegradedWhenInPendingPhaseFor(monitorPod *corev1.Pod, maxPendingDuration time.Duration, startupPodDegraded *operatorv1.OperatorCondition) {
+	if monitorPod.Status.Phase != corev1.PodPending || monitorPod.Status.StartTime == nil || time.Now().Sub(monitorPod.Status.StartTime.Time) < maxPendingDuration {
```
Comment (sttts) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:79]: nit add a oneliner comment above each block to make reading the flow easy

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+	startupPodDegraded.Status = operatorv1.ConditionTrue
+	pendingPodReason, pendingPodMessage := "Unknown", "unknown"
+	if len(monitorPod.Status.Reason) > 0 {
+		pendingPodReason = monitorPod.Status.Reason
+	}
+	if len(monitorPod.Status.Message) > 0 {
+		pendingPodMessage = monitorPod.Status.Message
+	}
+	startupPodDegraded.Reason = pendingPodReason
+	startupPodDegraded.Message = fmt.Sprintf("the pod %v has been in %v phase for %v due to %v, the max tolerated value is %v", monitorPod.Name, corev1.PodPending, time.Now().Sub(monitorPod.Status.StartTime.Time), pendingPodMessage, maxPendingDuration)
```
Comment (sttts) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:92]: don't use durations in messages. Messages must be constant.

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+	monitorPod, err := fd.podLister.Get(fmt.Sprintf("%s-startup-monitor", fd.targetName))
+	if errors.IsNotFound(err) {
+		return nil
+	} else if err != nil {
+		return err // unknown error
+	}
+
+	withDegradedWhenInPendingPhaseFor(monitorPod, 2*time.Minute, startupPodDegraded)
+	withDegradedWhenHasFailedContainer(monitorPod, startupPodDegraded)
+	withDegradedWhenHasExcessiveRestarts(monitorPod, 2, startupPodContainerExcessiveRestartsDegraded)
```
Comment (aojea) [inline pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:73]: is it possible that `withDegradedWhenHasExcessiveRestarts` shadows some of the previous?

Comment (aojea) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:73]: I mean, should any of `withDegradedWhen*`return immediately once if finds a problem?

Comment (p0lyn0mial) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:73]: I don't think so, hence a new condition. The previous condition will be set when a pod failed with a non-zero exit code or has been pending for more than 2 minutes.

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go (hunk 3 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:52]: I am not aware of any.

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go (hunk 5 above)
Comment (aojea) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:92]: actually, the time in the message will be different than the one used in the `if`

File: pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go (hunk 4 above)
Comment (p0lyn0mial) [reply pkg/operator/staticpod/controller/startupmonitorcondition/startup_monitor_condition.go:84]: done

**Pull request merged**